client/scala/src/main/scala/ai/verta/blobs/dataset/Dataset.scala

```diff
@@ -5,14 +5,110 @@ import ai.verta.blobs._
+  /** TODO: Figure out a way to remove this */
+  // Function to downwload a component, given its path in the blob, the blob's path in the commit
+  // and the file pointing to local path to download to:
+  private[verta] var downloadFunction: Option[(String, String, File) => Try[Unit]] = None
+  private[verta] var blobPath: Option[String] = None // path to the blob in the commit
+
+  /** Downloads componentPath from this dataset if ModelDB-managed versioning was enabled
+   *  Currently, only support downloading to a specific path
+   *  @param componentPath Original path of the file or directory in this dataset to download
+   *  @param downloadToPath Path to download to
+   *  @return Whether the download attempts succeed.
+   */
+  def download(
+    componentPath: Option[String] = None,
+    downloadToPath: String
+  )(implicit ec: ExecutionContext): Try[Unit] = {
+    /** TODO: Make downloadToPath optional */
+
+    if (!enableMDBVersioning)
+      Failure(new IllegalStateException("This blob did not allow for versioning"))
+    else if (downloadFunction.isEmpty || blobPath.isEmpty)
+      Failure(new IllegalStateException(
+        "This dataset cannot be used for downloads. Consider using `commit.get()` to obtain a download-capable dataset"
+      ))
+    else {
+      val componentToLocalPath = determineComponentAndLocalPaths(componentPath, downloadToPath)
+
+      if (componentToLocalPath.isEmpty)
+        Failure(new NoSuchElementException("Components not found."))
+      else
+        Try {
+          componentToLocalPath
+            .map(pair => downloadComponent(pair._1, pair._2))
+            .map(_.get)
+        }
+    }
+  }
+
+  /** Download a single component, to a determined local destination
+   *  @param componentPath Path to the component
+   *  @param downloadToPath Local path to download to
+   *  @return whether the download attempt succeeds.
+   */
+  private def downloadComponent(
+    componentPath: String,
+    downloadToPath: String
+  )(implicit ec: ExecutionContext): Try[Unit] = {
+    val file = new File(downloadToPath)
+
+    Try ({
+      Option(file.getParentFile()).map(_.mkdirs()) // create the ancestor directories, if necessary
+      file.createNewFile() // create the new file, if necessary
+    })
+      .flatMap(_ => downloadFunction.get(blobPath.get, componentPath, file))
+  }
+
+  /** Identify components to be downloaded, along with their local destination paths.
+   *  @param componentPath (Optional) path to directory or file within blob.
+   *  @param downloadToPath Local path to download to
+   *  @return Map of component paths to local destination paths
+   */
+  private def determineComponentAndLocalPaths(
+    componentPath: Option[String] = None,
+    downloadToPath: String
+  ): Map[String, String] = {
+    if (componentPath.isEmpty) {
+      // download entire blob
+      val downloadToPaths =
+        listPaths.map(comp => joinPaths(downloadToPath, removePrefixDir(comp, "s3:")))
```

Comment:
Does this match what the client does?

Reply:
I believe so. This is the corresponding section (according to my understanding of Michael's implementation):
https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/dataset/_dataset.py#L138